### PR TITLE
[in_app_purchase_storekit] Add storekit 2 support for canMakePayments and products

### PR DESCRIPTION
**Pull request by LouiseHsu**
This PR contains support for StoreKit2's [canMakePayments](https://developer.apple.com/documentation/storekit/appstore/3822277-canmakepayments) and [products](https://developer.apple.com/documentation/storekit/product/3851116-products).

This also contains basic scaffolding for SK2 support, such as pigeon translators.

Part of https://github.com/flutter/flutter/issues/116383

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

### CODE REVIEW

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.swift [hunk 1]
```diff
@@ -41,6 +41,9 @@ public class InAppPurchasePlugin: NSObject, FlutterPlugin, InAppPurchaseAPI {
     registrar.addMethodCallDelegate(instance, channel: channel)
     registrar.addApplicationDelegate(instance)
     SetUpInAppPurchaseAPI(messenger, instance)
+    if #available(iOS 15.0, *) {
+      InAppPurchase2APISetup.setUp(binaryMessenger: messenger, api: instance)
```
Comment (hellohuanlin) [inline packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.swift:45]: `InAppPurchaseStoreKit2APISetup`

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/StoreKit2/InAppPurchasePlugin2.swift [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  // https://developer.apple.com/documentation/storekit/product/3851116-products
+  func products(
+    identifiers: [String], completion: @escaping (Result<[SK2ProductMessage], any Error>) -> Void
+  ) {
+    Task {
+      do {
+        let products = try await Product.products(for: identifiers)
+        let productMessages = products.map { product in
+          product.convertToPigeon()
+        }
```
Comment (hellohuanlin) [inline packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/StoreKit2/InAppPurchasePlugin2.swift:25]: `products.map { $0.convertToPigeon() }`

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/StoreKit2/StoreKit2Translators.swift [hunk 3]
```diff
@@ -0,0 +1,169 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import Foundation
+import StoreKit
+
+@available(iOS 15.0, macOS 12.0, *)
+extension Product {
+  func convertToPigeon() -> SK2ProductMessage {
```
Comment (hellohuanlin) [inline packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/StoreKit2/StoreKit2Translators.swift:10]:
```
var pigeonMessage: SK2ProductMessage {
  return ...
}
```

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/StoreKit2/StoreKit2Translators.swift [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    return lhs.id == rhs.id && lhs.displayName == rhs.displayName
+      && lhs.description == rhs.description && lhs.price == rhs.price
+      && lhs.displayPrice == rhs.displayPrice && lhs.type == rhs.type
+      && lhs.subscription == rhs.subscription && lhs.priceLocale == rhs.priceLocale
+  }
+}
+
+@available(iOS 15.0, macOS 12.0, *)
+extension Product.ProductType {
+  func convertToPigeon() -> SK2ProductTypeMessage {
```
Comment (hellohuanlin) [inline packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/StoreKit2/StoreKit2Translators.swift:36]: same here

File: packages/in_app_purchase/in_app_purchase_storekit/example/ios/RunnerTests/InAppPurchase2PluginTests.swift [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+import StoreKitTest
+import XCTest
+
+@testable import in_app_purchase_storekit
+
+@available(iOS 15.0, *)
+final class InAppPurchase2PluginTests: XCTestCase {
+  var session: SKTestSession!
+  var plugin: InAppPurchasePlugin!
```
Comment (hellohuanlin) [inline packages/in_app_purchase/in_app_purchase_storekit/example/ios/RunnerTests/InAppPurchase2PluginTests.swift:13]: can be private

File: packages/in_app_purchase/in_app_purchase_storekit/example/ios/RunnerTests/InAppPurchase2PluginTests.swift [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+      case .failure(_):
+        XCTFail("Products should be successfully fetched")
+      }
+    }
+    await fulfillment(of: [expectation], timeout: 5)
+
+    XCTAssert(fetchedProductMsg?.count == 0)
+  }
+
+  @available(iOS 17.0, *)
```
Comment (hellohuanlin) [inline packages/in_app_purchase/in_app_purchase_storekit/example/ios/RunnerTests/InAppPurchase2PluginTests.swift:73]: add a todo to support testing on lower ios version

File: packages/in_app_purchase/in_app_purchase_storekit/lib/src/in_app_purchase_storekit_platform.dart [hunk 7]
```diff
@@ -17,6 +18,9 @@ const String kPurchaseErrorCode = 'purchase_error';
 /// Indicates store front is Apple AppStore.
 const String kIAPSource = 'app_store';
 
+/// Experimental flag for StoreKit2.
+bool _useStoreKit2 = false;
```
Comment (hellohuanlin) [inline packages/in_app_purchase/in_app_purchase_storekit/lib/src/in_app_purchase_storekit_platform.dart:22]: move inside the class

File: packages/in_app_purchase/in_app_purchase_storekit/lib/src/types/app_store_product_details.dart [hunk 8]
```diff
@@ -5,6 +5,7 @@
 import 'package:in_app_purchase_platform_interface/in_app_purchase_platform_interface.dart';
 
 import '../../store_kit_wrappers.dart';
+import '../store_kit_2_wrappers/sk2_product_wrapper.dart';
 
 /// The class represents the information of a product as registered in the Apple
 /// AppStore.
```
Comment (hellohuanlin) [inline packages/in_app_purchase/in_app_purchase_storekit/lib/src/types/app_store_product_details.dart:11]: can you update comments

**Review by hellohuanlin — APPROVED** [review]
LGTM, just a few comments

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/StoreKit2/StoreKit2Translators.swift [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+      price: NSDecimalNumber(decimal: price).doubleValue,
+      displayPrice: displayPrice,
+      type: type.convertToPigeon(),
+      subscription: subscription?.convertToPigeon(),
+      priceLocale: priceFormatStyle.locale.convertToPigeon()
+    )
+  }
+}
+
+extension SK2ProductMessage: Equatable {
```
Comment (hellohuanlin) [inline packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/StoreKit2/StoreKit2Translators.swift:25]: try remove `==`

File: packages/in_app_purchase/in_app_purchase_storekit/lib/src/types/app_store_product_details.dart (hunk 8 above)
Comment (LouiseHsu) [reply packages/in_app_purchase/in_app_purchase_storekit/lib/src/types/app_store_product_details.dart:11]: done

File: packages/in_app_purchase/in_app_purchase_storekit/lib/src/in_app_purchase_storekit_platform.dart (hunk 7 above)
Comment (LouiseHsu) [reply packages/in_app_purchase/in_app_purchase_storekit/lib/src/in_app_purchase_storekit_platform.dart]: done

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/StoreKit2/StoreKit2Translators.swift (hunk 4 above)
Comment (LouiseHsu) [reply packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/StoreKit2/StoreKit2Translators.swift]: done

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/StoreKit2/StoreKit2Translators.swift (hunk 3 above)
Comment (LouiseHsu) [reply packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/StoreKit2/StoreKit2Translators.swift]: done

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/StoreKit2/InAppPurchasePlugin2.swift (hunk 2 above)
Comment (LouiseHsu) [reply packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/StoreKit2/InAppPurchasePlugin2.swift]: done

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.swift (hunk 1 above)
Comment (LouiseHsu) [reply packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.swift:45]: done

File: packages/in_app_purchase/in_app_purchase_storekit/example/ios/RunnerTests/InAppPurchase2PluginTests.swift (hunk 5 above)
Comment (LouiseHsu) [reply packages/in_app_purchase/in_app_purchase_storekit/example/ios/RunnerTests/InAppPurchase2PluginTests.swift]: done

File: packages/in_app_purchase/in_app_purchase_storekit/example/ios/RunnerTests/InAppPurchase2PluginTests.swift (hunk 6 above)
Comment (LouiseHsu) [reply packages/in_app_purchase/in_app_purchase_storekit/example/ios/RunnerTests/InAppPurchase2PluginTests.swift]: done

**Pull request merged**